[mlir][LLVM] Add the `ConvertToLLVMAttrInterface` and `ConvertToLLVMOpInterface` interfaces

fabianmcg opened the pull request:
This patch adds the `ConvertToLLVMAttrInterface` and `ConvertToLLVMOpInterface` interfaces. It also modifies the `convert-to-llvm` pass to use these interfaces when available.

The `ConvertToLLVMAttrInterface` interfaces allows attributes to configure conversion to LLVM, including the conversion target, LLVM type converter, and populating conversion patterns. See the `NVVMTargetAttr` implementation of this interface for an example of how this interface can be used to configure conversion to LLVM.

The `ConvertToLLVMOpInterface` interface collects all convert to LLVM attributes stored in an operation.

Finally, the `convert-to-llvm` pass was modified to use these interfaces when available. This allows applying `convert-to-llvm` to GPU modules and letting the `NVVMTargetAttr` decide which patterns to populate.

fabianmcg commented:
The reasoning behind naming them `TransformsInterfaces` is because I intend in future patches add interfaces for:
- Configuring a pass manager
- Populate regular pattern rewrites

Also, in future patches I'll add the interface for `ROCDL`, `NVGPU`, ....

joker-eph commented on mlir/include/mlir/Interfaces/TransformsInterfaces.td:
```diff
@@ -0,0 +1,77 @@
+      /*args=*/(ins "::mlir::ConversionPatternAttrOptions&":$options,
+                    "::mlir::RewritePatternSet&":$patternSet)>
+  ];
+}
```
I don't understand what this interface provides? 
"conversion" is very context dependent: for example a math operation can have different conversion for going to LLVM, to SPIRV, or to TensorFlow for example.

fabianmcg replied:
It provides the ability to configure the dialect conversion data structures, regardless if the target is LLVM, SPIR-V, or things like `scf` to `cf`.

I debated with myself whether scope it to a `ConvertToLLVM` interface, or let it have a more generic purpose if someone else found it useful, and avoid recreating this interface for every possible dialect conversion.

I choose the generic version. However, I don't have strong opinions for it, I can switch it to a `ConvertToLLVM` specific interface.

joker-eph replied:
> regardless if the target is LLVM, SPIR-V, or things like scf to cf. 

Right: I don't quite understand this. How can it be useful to have this interface if you don't know the target? The patterns to populate for an operation are entirely dependent of the conversion you're doing aren't they?

fabianmcg replied:
Yes.
That's precisely why I pass `ConversionPatternAttrOptions` by reference, to sub-class it and allow interfaces to discern which conversion to populate. But also at the same time, not forcing every conversion to add an interface for it. See for example the `NVVMTargetPatternsAttrInterface::populateConversionPatterns`:
```c++
void NVVMTargetPatternsAttrInterface::populateConversionPatterns(
    Attribute attr, ConversionPatternAttrOptions &options,
    RewritePatternSet &patterns) const {
  auto *llvmOptions = dyn_cast<LLVMConversionPatternAttrOptions>(&options);
  // Bail if the options are invalid.
  if (!llvmOptions)
    return;
  configureGpuToNVVMConversionLegality(options.getConversionTarget());
  configureGpuToNVVMTypeConverter(llvmOptions->getLLVMTypeConverter());
  populateGpuToNVVMConversionPatterns(llvmOptions->getLLVMTypeConverter(),
                                      patterns);
}
```
However, I see 100% the point of having interfaces for each conversion. I can change it to an LLVM specific one.

matthias-springer commented on mlir/lib/Conversion/GPUToNVVM/LowerGpuOpsToNVVMOps.cpp:
```diff
@@ -409,3 +415,38 @@ void mlir::populateGpuToNVVMConversionPatterns(LLVMTypeConverter &converter,
+};
+} // namespace
+
+void NVVMTargetPatternsAttrInterface::populateConversionPatterns(
```
This implementation does not use `attr`. What's the benefit of this attr interface over the existing `ConvertToLLVMPatternInterface` dialect interface, which seems to do almost the same thing? Do you have a use kind in mind that cannot be solved with `ConvertToLLVMPatternInterface`?

fabianmcg replied:
> This implementation does not use attr.

That's something that should change in future patches. Right now I tried to keep the patch small.

> What's the benefit of this attr interface over the existing ConvertToLLVMPatternInterface dialect interface, which seems to do almost the same thing

`ConvertToLLVMPatternInterface` is a dialect interface, hence it is possible to use it for things like `arith-to-llvm`, because the lowering is mostly target independent -it shouldn't be but that's a different discussion.

However, that approach doesn't work for the `gpu` dialect because `gpu` goes to either `nvvm` or `rocdl`, instead of `llvm`. Hence, implementing `ConvertToLLVMPatternInterface` for the `gpu` dialect is not possible with a dedicated cost-model to select the right patterns (either `nvvm` or `rocdl`).

This proposal solves the issue by adding an attribute interface allowing for finer configuration of the conversion process. For example, this interface allows converting things like:

```mlir
gpu.module @nvvm [#nvvm.target] {
 // uses NVVM patterns
 ...
}
gpu.module @nvvm [#nvvm.target<chip = "sm_90">] {
 // uses NVVM patterns specific to `sm_90`
 ...
}
gpu.module @rocdl [#rocdl.target] {
 // uses ROCDL patterns
 ...
}
```

fabianmcg committed:
[mlir][LLVM] Add the `ConvertToLLVMAttrInterface` and `ConvertToLLVMOpInterface` interfaces

This patch adds the `ConvertToLLVMAttrInterface` and `ConvertToLLVMOpInterface` interfaces. It also modifies the `convert-to-llvm` pass to use these interfaces when available.

The `ConvertToLLVMAttrInterface` interfaces allows attributes to configure conversion to LLVM, including the conversion target, LLVM type converter, and populating conversion patterns. See the `NVVMTargetAttr` implementation of this interface for an example of how this interface can be used to configure conversion to LLVM.

The `ConvertToLLVMOpInterface` interface collects all convert to LLVM attributes stored in an operation.

Finally, the `convert-to-llvm` pass was modified to use these interfaces when available. This allows applying `convert-to-llvm` to GPU modules and letting the `NVVMTargetAttr` decide which patterns to populate.

joker-eph commented on mlir/lib/Conversion/ConvertToLLVM/ConvertToLLVMPass.cpp:
```diff
@@ -104,20 +99,33 @@ class ConvertToLLVMPass
+    RewritePatternSet patterns(context);
+    ConversionTarget target(*context);
+    target.addLegalDialect<LLVM::LLVMDialect>();
+    LLVMTypeConverter typeConverter(context);
```
I'm concerned about the cost of all this at the moment :(

fabianmcg replied:
I can profile it and check the impact.

However, let's say the cost of that initialization is `C`. In the previous version `C` is paid only once.
In this version the overall cost is `C` * `threads`, so the overall computation time does increment.
However, on a thread per thread basis the cost remains relatively the same.

So, I would argue, that unless the effect of the change is severe, the extra overhead might be justified for the feature.

fabianmcg replied:
Test results:
LLVM Release with assertions off 
cpu: Intel(R) Xeon(R) Gold 6132 CPU @ 2.60GHz
Test: run 100 times `mlir-opt --pass-pipeline="builtin.module(convert-to-llvm)"  mlir/test/Conversion/FuncToLLVM/func-to-llvm.mlir --mlir-timing` 

convert-to-llvm time:
```
Pre change: 0.00210028567 s
Post change: 0.0019316432 s
Difference: 0.00016864247 s
Slowdown(post/pre): 1.08730518658932
```

Total time:
```
Pre change: 0.00816961453 s
Post change: 0.00827812259 s
Difference: 0.00010850806 s
Slowdown(post/pre): 1.01328190694941
```

Globally the impact seems to be minimal (less than 2%), per pass it seems to increase around 10%. However, the per pass time doesn't include initialization.

joker-eph replied:
> However, let's say the cost of that initialization is C. In the previous version C is paid only once. In this version the overall cost is C * threads, so the overall computation time does increase. However, on a thread per thread basis the cost might not be as impactful.

I believe the cost is `C * thread * N` where N is the number of invocation of the pass-pipeline: if I setup a compilation server my pass-pipeline could be initialized once and used to compiled many modules.

fabianmcg replied:
I also thought that could be the case. However, it appears that `initialize` is called every time. I added the pass multiple times in a pipeline and added a print in the initialize method of the method,  and the method is called every time. Seems like something to fix in the `PassManager`.

But even then, let's assume that the cost of one vs the other is `thread * N` times greater. If `convert-to-llvm` was a pass like `cse` or `canonicalize` I would 100% agree with you, however, this is a pass that likely won't be invoked that many times.

fabianmcg replied:
I can think 2 solutions to this issue with little to no performance impact. 
Create a new `convert-to-llvm` like pass, where we say the performance cost doesn't matter. 
Make `ConversionTarget`, `TypeConverter` and `RewritePatternSet` somewhat copyable.

joker-eph commented on mlir/lib/Conversion/ConvertToLLVM/ToLLVMInterface.cpp:
```diff
@@ -30,3 +30,20 @@ void mlir::populateConversionTargetFromOperation(
+  for (ConvertToLLVMAttrInterface attr : attrs)
+    attr.populateConvertToLLVMConversionPatterns(target, typeConverter,
+                                                 patterns);
+}
```
This seems a bit "ad-hoc" to me, and also it means that running the pass on a a module with such an attribute, or on a function inside the module would give a different result?

fabianmcg replied:
I think there's a missing `CompilationUnitOpInterface` abstraction in general in MLIR, which would clarify the semantics in this case, because one could search the nearest op with the interface (or trait).

However, I also think it is reasonable for users to expect that a pass might perform different actions depending on the op context. In the sense that `convert-to-llvm` applied to a function won't update `func.func`, but if applied to `module` containing `func.func` it will be converted to `llvm.func`.

joker-eph replied:
> However, I also think it is reasonable for users to expect that a pass might perform different actions depending on the op context.

Yes but here it's not the "op context": what I'm describing is the exact same input IR, but the pass pipeline configured differently

fabianmcg replied:
How about if the anchor interface is changed to:

```c++
ConvertToLLVMOpInterface getConvertToLLVMOpInterface(Operation* op) {
  if (auto iface = dyn_cast<ConvertToLLVMOpInterface>(op))
    return iface;
  return op->getParentOfType<ConvertToLLVMOpInterface>();
}
// ... ConvertToLLVMPass::runOnOperation()
  ConvertToLLVMOpInterface iface = getConvertToLLVMOpInterface(getOperation());
  ...
```

In that case, anything inside of a `ConvertToLLVMOpInterface` would get converted according to a target.

fabianmcg committed:
add option to control whether to use conversion attributes

fabianmcg commented:
@joker-eph I added an option to enable or disable using conversion attributes, making the performance cost optional.
I also updated `populateOpConvertToLLVMConversionPatterns` to check the op or parent ops for the conversion op interface.

I also conducted several other tests.
## First test
The first test was about the overhead of the change.
The commands used for testing were:
```
// Use the slow path
mlir-opt --pass-pipeline="builtin.module(convert-to-llvm{use-conversion-attrs=true})"  mlir/test/Conversion/FuncToLLVM/func-to-llvm.mlir --mlir-timing
// Use the fast path
mlir-opt --pass-pipeline="builtin.module(convert-to-llvm{use-conversion-attrs=false})"  mlir/test/Conversion/FuncToLLVM/func-to-llvm.mlir --mlir-timing
// Upstream
mlir-opt --pass-pipeline="builtin.module(convert-to-llvm)"  mlir/test/Conversion/FuncToLLVM/func-to-llvm.mlir --mlir-timing
```

Each of the commands were run 1000 times, under the following settings:
```
Build: LLVM Release with assertions off
CPU: Intel(R) Xeon(R) Gold 6132 CPU @ 2.60GHz
```

The results were:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<body link="#467886" vlink="#96607D">

  | C1: use-conversion-attrs=true | C2: use-conversion-attrs=false | C3: upstream | C1-C2 | C2-C3 | C1-C3
-- | -- | -- | -- | -- | -- | --
Pass   Manager: (s) | 0.002824224 | 0.002724727 | 0.002791926 | 9.94975E-05 | -6.71990E-05 | 3.22985E-05
ConvertToLLVM:   (s) | 0.002065047 | 0.001913615 | 0.001976492 | 1.51432E-04 | -6.28765E-05 | 8.85554E-05
Total:   (s) | 0.00545438 | 0.005311841 | 0.005429315 | 1.42539E-04 | -1.17474E-04 | 2.50649E-05
</body>
</html>

The pass manager times correspond to the times obtained by modifying `MlirOptMain` with:
```c++
  auto start = std::chrono::high_resolution_clock::now();
  if (failed(pm.run(*op)))
    return failure();
  auto stop = std::chrono::high_resolution_clock::now();
  auto duration = std::chrono::duration_cast<std::chrono::duration<double>>(stop -start);
```

These results show that the difference in performance between upstream and the change with `use-conversion-attrs=false` is nonexistent.  They also show that under the current testing conditions the performance incurred by `use-conversion-attrs=true` is almost nonexistent.


## Second test
I added the following lines to `ConvertToLLVMPass::initialize`:

```c++
  static int x = 0;
  llvm::errs() << "init: " << x++ << "\n";
```

Then run the pipeline `builtin.module(convert-to-llvm, convert-to-llvm)`, using the following C++ code:
```c++
  if (failed(pm.run(*op)))
    return failure();
  if (failed(pm.run(*op)))
    return failure();
```

The output was:
```
init: 0
init: 1
init: 2
init: 3
```

This means that the `PassManager` is not caching `Pass::initialize` . Hence, this change introduces no overhead under the current conditions (even with ` use-conversion-attrs=true`).

One final thing I discovered is that applying `builtin.module(func.func(convert-to-llvm))` segfaults because the pass attempts to convert `func.func` . Hence, the pass can only applied at the "module" level.

joker-eph commented on mlir/lib/Conversion/ConvertToLLVM/ConvertToLLVMPass.cpp:
```diff
@@ -61,6 +61,8 @@ class LoadDependentDialectExtension : public DialectExtensionBase {
+      interfaces;
   std::shared_ptr<const FrozenRewritePatternSet> patterns;
   std::shared_ptr<const ConversionTarget> target;
   std::shared_ptr<const LLVMTypeConverter> typeConverter;
```
There should be much more documentation around this, because you're introducing subtle invariants where fields are only used in one of two modes, and uninitialized in the other. This kind of modal behavior isn't great in general.


Can we actually refactor this by encapsulating this into another abstract class, with two implementations and then unifying all the downstream code from there?

fabianmcg replied:
Done, I created to sub classes `StaticConvertToLLVM` and `DynamicConvertToLLVM` controlled by the  `convert-to-llvm` `dynamic` flag.

fabianmcg committed:
address reviewer comments

fabianmcg committed:
fix shared lib build

joker-eph reviewed (approved):
I suspect we'll want to simplify this in the future, not clear to me the pass-option is desirable, but that looks good for now!

fabianmcg committed:
privatize the ConvertToLLVMPassInterface

fabianmcg commented:
> I suspect we'll want to simplify this in the future, not clear to me the pass-option is desirable

Yes, my next step is to add attributes to deduce the `LowerToLLVMOptions` options from the IR to include it in the dynamic version of the pass.
